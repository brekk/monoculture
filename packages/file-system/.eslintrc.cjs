module.exports = {
  root: true,
  extends: ['monoculture'],
  ignorePatterns: ['file-system.js'],
}
