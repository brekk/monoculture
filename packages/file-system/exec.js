import { pipe, curry, prop, ifElse, propOr, __ as $ } from 'ramda'
import { execa } from 'execa'
import { Future } from 'fluture'

export const didFail = propOr(true, 'failed')
export const fail = prop('stderr')
/* eslint-disable max-len */
/**
 * @pageSummary Call external processes and consume them as a Future-wrapped value.
 * This is a light wrapper around `execa`, so please see [execa](https://github.com/sindresorhus/execa)'s documentation for more info.
 */

/* eslint-disable jsdoc/tag-lines */
/**
 * Consume external commands as a Future-wrapped value.
 * @curried
 *  1. execWithConfig - Passes all possible configuration values plus a cancellation function.
 *
 *     @example
 *     ```js
 *     import { execWithConfig } from 'file-system'
 *     import { fork } from 'fluture'
 *     fork(console.warn)(console.log)(
 *       execWithConfig(
 *         function customCancellationFunction() {},
 *         'echo',
 *         { cleanup: true },
 *         ['ahoy']
 *       )
 *     )
 *     ```
 *
 *  2. execWithCancel - Eschews any configuration and instead only expects a cancellation function, command and arguments.
 *
 *     @example
 *     ```js
 *     import { execWithCancel } from 'file-system'
 *     import { fork } from 'fluture'
 *     fork(console.warn)(console.log)(
 *       execWithCancel(
 *         function customCancellationFunction() {},
 *         'echo',
 *         ['ahoy']
 *       )
 *     )
 *     ```
 *  3. exec - Eschews any configuration or cancellation function. Needs only command and arguments.
 *
 *     @example
 *     ```js
 *     import { exec } from 'file-system'
 *     import { fork } from 'fluture'
 *     fork(console.warn)(console.log)(
 *       exec(
 *         'echo',
 *         ['ahoy']
 *       )
 *     )
 *     ```
 */
/* eslint-enable max-len */
export const execWithConfig = curry(
  (cancellation, cmd, opts, args) =>
    new Future((bad, good) => {
      execa(cmd, args, opts)
        .catch(pipe(fail, bad))
        .then(ifElse(didFail, pipe(fail, bad), good))
      return cancellation
    })
)
export const execWithCancel = execWithConfig($, $, undefined)
export const exec = execWithCancel(() => {})
