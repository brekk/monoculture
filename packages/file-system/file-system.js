#!/usr/bin/env node

// src/flexeca.js
import { pipe, curry, ifElse, propOr } from "ramda";
import { execa } from "execa";
import { Future } from "fluture";
var didFail = propOr(true, "failed");
var fail = propOr("Something broke", "stderr");
var flexecaWithCanceller = curry(
  (cancellation, cmd, args) => new Future((bad, good) => {
    execa(cmd, args).catch(pipe(fail, (z) => `FAILED: "${z}"`, bad)).then(ifElse(didFail, pipe(fail, bad), good));
    return cancellation;
  })
);
var flexeca = flexecaWithCanceller(() => {
});

// src/fs.js
import {
  rm as __rm,
  mkdir as __mkdir,
  readFile as __readFile,
  writeFile as __writeFile,
  access as __access,
  constants
} from "node:fs";
import { propOr as propOr2, without, curry as curry2, pipe as pipe2, map } from "ramda";
import { chain, chainRej, Future as Future2, parallel } from "fluture";
import glob from "glob";
var localize = (z) => `./${z}`;
var readFile = (x) => new Future2((bad, good) => {
  __readFile(x, "utf8", (err, data) => err ? bad(err) : good(data));
  return () => {
  };
});
var readJSONFile = pipe2(readFile, map(JSON.parse));
var readDirWithConfig = curry2(
  (conf, g) => Future2((bad, good) => {
    try {
      glob(
        g,
        conf,
        (e, x) => (
          // thus far I cannot seem to ever call `bad` from within here
          e ? bad(e) : good(x)
        )
      );
    } catch (e) {
      bad(e);
    }
    return () => {
    };
  })
);
var readDir = readDirWithConfig({});
var writeFileWithConfig = curry2(
  (conf, file, content) => new Future2((bad, good) => {
    __writeFile(file, content, conf, (e) => {
      if (e) {
        bad(e);
        return;
      }
      good(content);
    });
    return () => {
    };
  })
);
var writeFile = writeFileWithConfig({ encoding: "utf8" });
var removeFileWithConfig = curry2(
  (options, fd) => new Future2((bad, good) => {
    __rm(fd, options, (err) => err ? bad(err) : good(fd));
    return () => {
    };
  })
);
var DEFAULT_REMOVAL_CONFIG = {
  force: false,
  maxRetries: 0,
  recursive: false,
  retryDelay: 100,
  parallel: 10
};
var removeFile = removeFileWithConfig(DEFAULT_REMOVAL_CONFIG);
var removeFilesWithConfig = curry2(
  (conf, list) => pipe2(
    map(removeFileWithConfig(without(["parallel"], conf))),
    parallel(propOr2(10, "parallel", conf))
  )(list)
);
var removeFiles = removeFilesWithConfig(DEFAULT_REMOVAL_CONFIG);
var mkdir = curry2(
  (conf, x) => new Future2((bad, good) => {
    __mkdir(x, conf, (err) => err ? bad(err) : good(x));
    return () => {
    };
  })
);
var mkdirp = mkdir({ recursive: true });
var access = curry2(
  (permissions, filePath) => new Future2((bad, good) => {
    __access(filePath, permissions, (err) => err ? bad(err) : good(true));
    return () => {
    };
  })
);
var exists = access(constants.F_OK);
var directoryOnly = (filePath) => filePath.slice(0, filePath.lastIndexOf("/"));
var writeFileWithAutoPath = curry2(
  (filePath, content) => pipe2(
    directoryOnly,
    (dir) => pipe2(
      exists,
      chainRej(() => mkdirp(dir))
    )(dir),
    chain(() => writeFile(filePath, content))
  )(filePath)
);
var rm = curry2(
  (conf, x) => new Future2((bad, good) => {
    __rm(x, conf, (err) => err ? bad(err) : good(x));
    return () => {
    };
  })
);
var rimraf = rm({ force: true, recursive: true });

// src/path.js
import { join, normalize } from "node:path";
import { curry as curry3 } from "ramda";
var pathRelativeTo = curry3((pwd, x) => {
  if (typeof pwd !== "string" || typeof x !== "string") {
    throw new Error("Cannot normalize bad paths.");
  }
  return join(pwd, normalize(x));
});

// src/interpret.js
import { Future as Future3 } from "fluture";
var interpret = (filepath) => Future3((bad, good) => {
  import(filepath).catch(bad).then(good);
  return () => {
  };
});
export {
  DEFAULT_REMOVAL_CONFIG,
  access,
  directoryOnly,
  exists,
  flexeca,
  flexecaWithCanceller,
  interpret,
  localize,
  mkdir,
  mkdirp,
  pathRelativeTo,
  readDir,
  readDirWithConfig,
  readFile,
  readJSONFile,
  removeFile,
  removeFileWithConfig,
  removeFiles,
  removeFilesWithConfig,
  rimraf,
  rm,
  writeFile,
  writeFileWithAutoPath,
  writeFileWithConfig
};
