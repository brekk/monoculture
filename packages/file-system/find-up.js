import { Future } from 'fluture'
import { curry } from 'ramda'
import { findUp as __findUp } from 'find-up'

export const findUpWithCancel = curry((cancel, opts, x) =>
  Future((bad, good) => {
    __findUp(x, opts)
      .catch(bad)
      .then(raw => (raw ? good(raw) : bad(new Error('No config file found!'))))
    return cancel
  })
)
export const findUp = findUpWithCancel(() => {})
