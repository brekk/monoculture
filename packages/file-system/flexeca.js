import { pipe, curry, ifElse, propOr } from 'ramda'
import { execa } from 'execa'
import { Future } from 'fluture'

const didFail = propOr(true, 'failed')
const fail = propOr('Something broke', 'stderr')

export const flexecaWithOptionsAndCancel = curry(
  (cancellation, cmd, opts, args) =>
    new Future((bad, good) => {
      execa(cmd, args, opts)
        .catch(pipe(fail, z => `FAILED: "${z}"`, bad))
        .then(ifElse(didFail, pipe(fail, bad), good))
      return cancellation
    })
)

export const flexecaWithCanceller = curry(
  (cancellation, cmd, args) =>
    new Future((bad, good) => {
      execa(cmd, args)
        .catch(pipe(fail, z => `FAILED: "${z}"`, bad))
        .then(ifElse(didFail, pipe(fail, bad), good))
      return cancellation
    })
)

export const flexeca = flexecaWithCanceller(() => {})
