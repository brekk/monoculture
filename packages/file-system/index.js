export * from './find-up'
export * from './exec'
export * from './fs'
export * from './interpret'
export * from './path'
