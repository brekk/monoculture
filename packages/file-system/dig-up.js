import { Future } from 'fluture'
import { curry } from 'ramda'
import { findUp as __findUp } from 'find-up'

export const digUpWithCancel = curry(
  function _digUpWithCancel(cancel, opts, x) {
    return Future((bad, good) => {
      __findUp(x, opts)
        .catch(bad)
        .then(raw =>
          raw ? good(raw) : bad(new Error('No config file found!'))
        )
      return cancel
    })
  }
)
export const digUp = digUpWithCancel(() => {})
