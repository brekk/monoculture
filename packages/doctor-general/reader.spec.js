import { fork, parallel, resolve as resolveF } from 'fluture'
import { chain, map, pipe } from 'ramda'
import { readFile } from 'file-system'
import {
  monorepoRunner,
  readPackageJsonWorkspaces,
  iterateOverWorkspacesAndReadFiles,
} from './reader'

test('iterateOverWorkspacesAndReadFiles', done => {
  fork(done)(x => {
    expect(x).toEqual([
      'tools/digested',
      'tools/doctor-general-cli',
      'tools/gitparty',
      'tools/spacework',
      'tools/superorganism',
      'tools/treacle',
    ])
    done()
  })(
    iterateOverWorkspacesAndReadFiles(
      {
        searchGlob: '*',
        ignore: [],
      },
      '../..',
      resolveF(['tools/'])
    )
  )
})

test('readPackageJsonWorkspaces', done => {
  pipe(
    readFile,
    map(JSON.parse),
    readPackageJsonWorkspaces('../..'),
    chain(parallel(10)),
    fork(done)(x => {
      expect(x).toMatchSnapshot()
      /*
      [
        ['apps/docs/'],
        [
          'packages/bloodline/',
          'packages/climate-json/',
          'packages/climate-toml/',
          'packages/climate-yaml/',
          'packages/climate/',
          'packages/clox/',
          'packages/doctor-general-jest/',
          'packages/doctor-general-mdx/',
          'packages/doctor-general/',
          'packages/file-system/',
          'packages/inherent/',
          'packages/kiddo/',
          'packages/knot/',
          'packages/manacle/',
          'packages/monocle/',
          'packages/monorail/',
          'packages/robot-tourist/',
          'packages/water-wheel/',
        ],
        [
          'shared/eslint-config-monoculture/',
          'shared/jest-config/',
          'shared/monoculture-tsconfig/',
        ],
        [
          'tools/digested/',
          'tools/doctor-general-cli/',
          'tools/gitparty/',
          'tools/spacework/',
          'tools/superorganism/',
          'tools/treacle/',
        ],
      ]
      */
      done()
    })
  )('../../package.json')
})

test('monorepoRunner', done => {
  const cancel = () => {
    done()
  }
  const config = {
    color: true,
    ignore: [
      '**/node_modules/**',
      '**/coverage/**',
      '**/*.spec.{js,jsx,ts,tsx}',
      '**/fixture/**',
      '**/fixture.*',
    ],
    searchGlob: '**/*.{js,jsx,ts,tsx}',
    debug: false,
    verifyProcessor: false,
    showMatchOnly: false,
  }
  fork(done)(x => {
    expect(x).toMatchSnapshot()
    done()
  })(monorepoRunner(cancel, config, '../..', '../../package.json'))
})
