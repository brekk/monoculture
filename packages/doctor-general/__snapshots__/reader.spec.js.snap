// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`monorepoRunner 1`] = `
[
  "apps/docs/blem.d.ts",
  "apps/docs/next-env.d.ts",
  "apps/docs/next.config.js",
  "apps/docs/pages/_app.tsx",
  "apps/docs/theme.config.tsx",
  "packages/bloodline/api.js",
  "packages/bloodline/bloodline.js",
  "packages/bloodline/cli.js",
  "packages/bloodline/config.js",
  "packages/bloodline/constants.js",
  "packages/bloodline/cyclic.js",
  "packages/bloodline/executables.js",
  "packages/bloodline/failure.js",
  "packages/bloodline/index.js",
  "packages/bloodline/log.js",
  "packages/bloodline/tree.js",
  "packages/bloodline/visualization.js",
  "packages/climate-json/climate-json.js",
  "packages/climate-toml/climate-toml.js",
  "packages/climate-yaml/climate-yaml.js",
  "packages/climate/builder.js",
  "packages/climate/climate.js",
  "packages/climate/events.js",
  "packages/climate/help.js",
  "packages/climate/log.js",
  "packages/climate/parser.js",
  "packages/clox/clox.js",
  "packages/doctor-general-jest/comment-test.js",
  "packages/doctor-general-jest/index.js",
  "packages/doctor-general-jest/log.js",
  "packages/doctor-general-jest/renderer-jest.js",
  "packages/doctor-general-mdx/comment-documentation.js",
  "packages/doctor-general-mdx/index.js",
  "packages/doctor-general-mdx/log.js",
  "packages/doctor-general-mdx/next-meta-files.js",
  "packages/doctor-general-mdx/renderer-markdown.js",
  "packages/doctor-general/comment.js",
  "packages/doctor-general/constants.js",
  "packages/doctor-general/doctor-general.js",
  "packages/doctor-general/drgen.js",
  "packages/doctor-general/file.js",
  "packages/doctor-general/log.js",
  "packages/doctor-general/parse.js",
  "packages/doctor-general/processor.js",
  "packages/doctor-general/reader.js",
  "packages/doctor-general/text.js",
  "packages/doctor-general/writer.js",
  "packages/file-system/dig-up.js",
  "packages/file-system/file-system.js",
  "packages/file-system/fs.js",
  "packages/file-system/future.js",
  "packages/file-system/interpret.js",
  "packages/file-system/log.js",
  "packages/file-system/path.js",
  "packages/file-system/test-helpers.js",
  "packages/inherent/array.js",
  "packages/inherent/common.js",
  "packages/inherent/groupByIndex.js",
  "packages/inherent/inherent.js",
  "packages/inherent/object.js",
  "packages/inherent/primitives.js",
  "packages/inherent/string.js",
  "packages/kiddo/kiddo.js",
  "packages/knot/constants.js",
  "packages/knot/knot.js",
  "packages/manacle/cli.js",
  "packages/manacle/config.js",
  "packages/manacle/examples/query-has-exports.js",
  "packages/manacle/log.js",
  "packages/monocle/cli.js",
  "packages/monocle/config.js",
  "packages/monocle/examples/plugin-c-words.js",
  "packages/monocle/examples/plugin-get-exports.js",
  "packages/monocle/examples/plugin-get-imported.js",
  "packages/monocle/examples/plugin-get-imports.js",
  "packages/monocle/examples/plugin-has-exports.js",
  "packages/monocle/examples/plugin-has-imports.js",
  "packages/monocle/examples/plugin-wordcount.js",
  "packages/monocle/hash.js",
  "packages/monocle/index.js",
  "packages/monocle/log.js",
  "packages/monocle/reader.js",
  "packages/monorail/helpers.js",
  "packages/monorail/list.js",
  "packages/monorail/log.js",
  "packages/monorail/monorail.js",
  "packages/monorail/runner.js",
  "packages/monorail/sort.js",
  "packages/monorail/validate.js",
  "packages/robot-tourist/cli.js",
  "packages/robot-tourist/config.js",
  "packages/robot-tourist/constants.js",
  "packages/robot-tourist/core.js",
  "packages/robot-tourist/index.js",
  "packages/robot-tourist/log.js",
  "packages/robot-tourist/plugin-robot-tourist.js",
  "packages/robot-tourist/plugin-simplifier.js",
  "packages/robot-tourist/regex.js",
  "packages/robot-tourist/reporter.js",
  "packages/robot-tourist/source-matcher.js",
  "packages/robot-tourist/stats.js",
  "packages/robot-tourist/string.js",
  "packages/robot-tourist/tuple.js",
  "packages/robot-tourist/wordbot.js",
  "packages/water-wheel/water-wheel.js",
  "shared/eslint-config-monoculture/index.js",
  "shared/jest-config/jest.config.js",
  "shared/jest-config/jest.setup.js",
  "shared/monoculture-tsconfig/test.ts",
  "tools/digested/config.js",
  "tools/digested/digested.js",
  "tools/digested/executable.js",
  "tools/digested/link.js",
  "tools/digested/log.js",
  "tools/digested/markdown.js",
  "tools/digested/summary.js",
  "tools/doctor-general-cli/cli.js",
  "tools/doctor-general-cli/config.js",
  "tools/doctor-general-cli/constants.js",
  "tools/doctor-general-cli/executable.js",
  "tools/doctor-general-cli/log.js",
  "tools/gitparty/dist/gitparty.js",
  "tools/gitparty/src/alias.js",
  "tools/gitparty/src/cli.js",
  "tools/gitparty/src/config.js",
  "tools/gitparty/src/errors.js",
  "tools/gitparty/src/filters.js",
  "tools/gitparty/src/git.js",
  "tools/gitparty/src/legend.js",
  "tools/gitparty/src/log.js",
  "tools/gitparty/src/per-commit.js",
  "tools/gitparty/src/runner.js",
  "tools/superorganism/package-scripts.js",
  "tools/superorganism/src/cli.js",
  "tools/superorganism/src/config.js",
  "tools/superorganism/src/helpers.js",
  "tools/superorganism/src/index.js",
  "tools/superorganism/src/log.js",
  "tools/superorganism/src/recursive.js",
  "tools/superorganism/src/runner.js",
  "tools/treacle/dist/treacle.js",
  "tools/treacle/src/cli.js",
  "tools/treacle/src/index.js",
]
`;

exports[`readPackageJsonWorkspaces 1`] = `
[
  [
    "apps/docs/",
  ],
  [
    "packages/bloodline/",
    "packages/climate-json/",
    "packages/climate-toml/",
    "packages/climate-yaml/",
    "packages/climate/",
    "packages/clox/",
    "packages/doctor-general-jest/",
    "packages/doctor-general-mdx/",
    "packages/doctor-general/",
    "packages/file-system/",
    "packages/inherent/",
    "packages/kiddo/",
    "packages/knot/",
    "packages/manacle/",
    "packages/monocle/",
    "packages/monorail/",
    "packages/robot-tourist/",
    "packages/water-wheel/",
  ],
  [
    "shared/eslint-config-monoculture/",
    "shared/jest-config/",
    "shared/monoculture-tsconfig/",
  ],
  [
    "tools/digested/",
    "tools/doctor-general-cli/",
    "tools/gitparty/",
    "tools/spacework/",
    "tools/superorganism/",
    "tools/treacle/",
  ],
]
`;
